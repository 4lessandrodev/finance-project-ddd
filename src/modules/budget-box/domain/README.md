# BudgetBox - Aggregate

---

São os caixas financeiros os quais o usuário irá cadastrar e definir sua meta em percentual.
Podendo também ser categorizado como um benefício.

Caso este esteja definido como benefício o percentual não pode ser diferente de 100%.

A soma de todos os caixas cadastrados e definidos como percentual não pode ultrapassar 100%.
bem como nenhum caixa percentual individual não pode ter a meta superior a 100%.

Cada caixa possui motivos que são pertinentes apenas a ele mesmo.

Os motivos tem como objetivo identificar a finalidade do lançamento.

A cada lançamento de entrada este receberá o montante de acordo com o percentual definido:

exemplo:

- Lançamento: Entrada;
- Valor: R$ 100;
- Percentual definido no caixa: 80;
- Montante a ser creditado no caixa: 80;

```json
{
  "id": "uuid",
  "owner-id": "uuid",
  "description": "valid_description",
  "balance-available": 1000,
  "is-percentual": true,
  "budget-percentage": 80,
  "reasons": [
    {
      "id": "uuid",
      "description": "valid_descripíont"
    }
  ]
}
```

### Structure

- Budgetbox: Aggregate - Ok
- reason-description: Value Object - Ok
- budget-description: Value Object - Ok
- budget-id: Value Object - Ok
- budget-percentage: Value Object - Ok
- reason: Entity - Ok
- reason-id: Value Object - Ok

## Fluxos
#### Excluir um caixa financeiro

<img src="../../../../readme/excluir-caixa.jpg" alt="imagem" width="100%">
