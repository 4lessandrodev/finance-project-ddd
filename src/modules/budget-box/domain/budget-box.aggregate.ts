import { AggregateRoot, BaseDomainEntity, CurrencyValueObject, DomainId, Result } from 'types-ddd';
import { BudgetDescriptionValueObject } from './budget-description.value-object';
import { ReasonDomainEntity } from './reason.domain-entity';
import {
	BUDGET_PERCENTAGE_MAX_VALUE,
	DEFAULT_BUDGET_PERCENTAGE_VALUE,
	PercentageValueObject
} from './percentage.value-object';
import ReasonDescriptionValueObject from './reason-description.value-object';
import BudgetBoxDeletedEvent from './events/budget-box-deleted.event';

export interface BudgetAggregateProps extends BaseDomainEntity {
	ownerId: DomainId;
	description: BudgetDescriptionValueObject;
	balanceAvailable: CurrencyValueObject;
	isPercentage: boolean;
	budgetPercentage: PercentageValueObject;
	reasons: ReasonDomainEntity[];
}
/**
	@var ownerId: `UserIdValueObject`
	@var description: `BudgetDescriptionValueObject`
	@var balanceAvailable: `CurrencyValueObject`
	@var isPercentage: `boolean`
	@var budgetPercentage: `PercentageValueObject`
	@var reasons: `ReasonDomainEntity[]`
 */
export class BudgetBoxAggregate extends AggregateRoot<BudgetAggregateProps> {
	private constructor (props: BudgetAggregateProps) {
		super(props, BudgetBoxAggregate.name);
	}

	get ownerId (): DomainId {
		return this.props.ownerId;
	}

	get description (): BudgetDescriptionValueObject {
		return this.props.description;
	}

	get balanceAvailable (): CurrencyValueObject {
		return this.props.balanceAvailable;
	}

	get isPercentage (): boolean {
		return this.props.isPercentage;
	}

	get budgetPercentage (): PercentageValueObject {
		return this.props.budgetPercentage;
	}

	get reasons (): ReasonDomainEntity[] {
		return this.props.reasons;
	}

	changePercentage (percentage: PercentageValueObject): void {
		if (!this.isPercentage) return;
		this.props.budgetPercentage = percentage;
	}

	addReason (reason: ReasonDomainEntity): void {
		this.props.reasons.push(reason);
	}

	getReasonById (reasonId: DomainId): ReasonDomainEntity | null {
		const reasonOrUndefined = this.props
			.reasons
			.find((reason) => reason
				.id.equals(reasonId));
		
		if (!reasonOrUndefined) {
			return null;
		}

		return reasonOrUndefined;
	}

	changeDescription (description: BudgetDescriptionValueObject): void {
		this.props.description = description;
	}

	changeReasonDescription (reasonId: DomainId, description: ReasonDescriptionValueObject): boolean {
		const updated = true;
		const reasonOrNull = this.getReasonById(reasonId);

		if (!reasonOrNull) {
			return !updated;
		}

		const reason = reasonOrNull.clone({
			idStrategy: 'uuid', isNew: false, props: {
				ID: reasonId,
				description,
				createdAt: reasonOrNull.createdAt
			}
		}).getResult();
		
		this.removeReasonById(reasonId);
		this.addReason(reason);
		return updated;
	}

	removeReasonById (reasonId: DomainId): boolean {
		const reasonIndex = this.props
			.reasons
			.findIndex((reason) => reason
				.id.equals(reasonId));
		
		const exists = reasonIndex !== -1;
		if (exists) {
			this.props.reasons.splice(reasonIndex, 1);	
		}

		return exists;
	}

	delete (): void {
		this.props.deletedAt = new Date();
		this.props.isDeleted = true;
		this.addDomainEvent(new BudgetBoxDeletedEvent(this));
	}

	public static create (
		props: BudgetAggregateProps
	): Result<BudgetBoxAggregate> {

		if (
			!props.isPercentage &&
			props.budgetPercentage.value !== BUDGET_PERCENTAGE_MAX_VALUE
		) {
			props.budgetPercentage = PercentageValueObject.create(
				DEFAULT_BUDGET_PERCENTAGE_VALUE,
			).getResult();
		}

		return Result.ok<BudgetBoxAggregate>(new BudgetBoxAggregate(props));
	}
}

export default BudgetBoxAggregate;
