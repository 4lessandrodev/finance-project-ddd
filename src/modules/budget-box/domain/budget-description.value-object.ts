
import { ErrorMessages } from '@modules/shared';
import { Result, ValueObject } from 'types-ddd';

export const BUDGET_DESCRIPTION_MAX_LENGTH = 30;
export const BUDGET_DESCRIPTION_MIN_LENGTH = 1;

export interface BudgetDescriptionValueObjectProps {
  value: string;
}

export class BudgetDescriptionValueObject extends ValueObject<BudgetDescriptionValueObjectProps> {
	private constructor (props: BudgetDescriptionValueObjectProps) {
		super(props);
	}

	get value (): string {
		return this.props.value;
	}

	static isValidValue (value: string): boolean {
		
		const descriptionLength = value.trim().length;

		const isValid = descriptionLength >= BUDGET_DESCRIPTION_MIN_LENGTH &&
			descriptionLength <= BUDGET_DESCRIPTION_MAX_LENGTH;
		return isValid;
	}

	public static create (
		description: string,
	): Result<BudgetDescriptionValueObject> {
		
		const isValidLength = BudgetDescriptionValueObject.isValidValue(description);
      
		if (!isValidLength) {
			return Result.fail<BudgetDescriptionValueObject>(
				ErrorMessages.INVALID_BUDGET_DESCRIPTION_LENGTH,
			);
		}
		
		return Result.ok<BudgetDescriptionValueObject>(
			new BudgetDescriptionValueObject({ value: description.toLowerCase() }),
		);
	}
}

export default BudgetDescriptionValueObject;
