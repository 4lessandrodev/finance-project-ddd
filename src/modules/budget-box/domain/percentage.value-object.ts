import { ErrorMessages } from '@shared/index';
import { ValueObject, Result } from 'types-ddd';

export const BUDGET_PERCENTAGE_MAX_VALUE = 100;
export const BUDGET_PERCENTAGE_MIN_VALUE = 0;
export const DEFAULT_BUDGET_PERCENTAGE_VALUE = 100;

export interface PercentageValueObjectProps {
  value: number;
}

export class PercentageValueObject extends ValueObject<PercentageValueObjectProps> {
	private constructor (props: PercentageValueObjectProps) {
		super(props);
	}

	get value (): number {
		return this.props.value;
	}

	static isValidValue (value: number): boolean {
		const isValidRange =
		value >= BUDGET_PERCENTAGE_MIN_VALUE &&
			value <= BUDGET_PERCENTAGE_MAX_VALUE;
		return isValidRange;
	}

	public static create (value: number): Result<PercentageValueObject> {

		const isValidRange = PercentageValueObject.isValidValue(value);

		if (!isValidRange) {
			return Result.fail<PercentageValueObject>(
				ErrorMessages.INVALID_PERCENTAGE_VALUE,
			);
		}

		return Result.ok<PercentageValueObject>(
			new PercentageValueObject({ value }),
		);
	}
}

export default PercentageValueObject;
