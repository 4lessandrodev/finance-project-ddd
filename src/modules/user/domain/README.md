# User - Aggregate

---

Este agregado identifica cada indivíduo cadastrado na plataforma, 
bem como controla o acesso e define o proprietário de cada registro.

```json
{
  "id": "uuid",
  "email": "valid_email@domain.com",
  "password": "valid_password",
  "createdAt": "2021-01-01 10:00:00",
  "updatedAt": "2021-01-01 10:00:00",
  "deletedAt": "",
  "terms": [
    {
      "ip": "127.0.0.1",
      "acceptedAt": "2021-01-01 10:00:00",
      "userAgent": {
        "name": "firefox",
        "version": "86.0",
        "os": "windows",
        "type": "browser",
      }
    },
  ]
}
```

- email: Value Object - Ok
- password: Value Object - Ok
- terms: Value Object - Ok
- ip: Value Object - Ok

### Fluxos

#### Excluir uma conta de usuário

<img src="../../../../readme/excluir-usuario.jpg" alt="imagem" width="100%">
