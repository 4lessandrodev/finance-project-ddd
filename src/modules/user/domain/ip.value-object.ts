import { ErrorMessages } from '@shared/index';
import { Result, ValueObject } from 'types-ddd';
const validateIpRegex = /\b(?:(?:2(?:[0-4][0-9]|5[0-5])|[0-1]?[0-9]?[0-9])\.){3}(?:(?:2([0-4][0-9]|5[0-5])|[0-1]?[0-9]?[0-9]))\b/;

export interface IpValueObjectProps {
  value: string;
}

export class IpValueObject extends ValueObject<IpValueObjectProps> {
	private constructor (props: IpValueObjectProps) {
		super(props);
	}

	get value (): string {
		return this.props.value;
	}

	public static isValidValue (ip: string): boolean {
		return validateIpRegex.test(ip);
	}

	public static create (ip: string): Result<IpValueObject> {
		
		const isValidIp = IpValueObject.isValidValue(ip);

		if (!isValidIp) {
			return Result.fail<IpValueObject>(ErrorMessages.INVALID_IP);
		}
		return Result.ok<IpValueObject>(new IpValueObject({ value: ip }));
	}
}

export default IpValueObject;
