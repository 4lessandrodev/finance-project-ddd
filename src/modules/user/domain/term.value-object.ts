import { ValueObject, Result, DateValueObject } from 'types-ddd';
import { IpValueObject } from './ip.value-object';

export interface IUserAgent {
	name: string;
	version: string;
	os: string;
	type: string;
}

export interface TermValueObjectProps {
	ip: IpValueObject;
	acceptedAt: DateValueObject;
	userAgent: IUserAgent;
	isAccepted: boolean;
}

export class TermValueObject extends ValueObject<TermValueObjectProps> {
	private constructor (props: TermValueObjectProps) {
		super(props);
	}

	get ip (): IpValueObject {
		return this.props.ip;
	}

	get acceptedAt (): DateValueObject {
		return this.props.acceptedAt;
	}

	get userAgent (): IUserAgent {
		return this.props.userAgent;
	}

	public static create (props: TermValueObjectProps): Result<TermValueObject> {
		if (!props.isAccepted) {
			return Result.fail('Terms must be accepted');
		}
		return Result.ok<TermValueObject>(new TermValueObject(props));
	}
}

export default TermValueObject;
