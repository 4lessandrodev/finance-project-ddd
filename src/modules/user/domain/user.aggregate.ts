import { TermValueObject } from './term.value-object';
import {
	AggregateRoot,
	BaseDomainEntity,
	EmailValueObject,
	PasswordValueObject,
	Result
} from 'types-ddd';
import UserAccountDeletedEvent from './events/delete-user-account.event';

export interface UserAggregateProps extends BaseDomainEntity {
	email: EmailValueObject;
	password: PasswordValueObject;
	terms: TermValueObject[];
}

/**
 * @var email: `EmailValueObject`
 * @var password: `PasswordValueObject`
 * @var terms: `TermValueObject[]`
 */
export class UserAggregate extends AggregateRoot<UserAggregateProps> {
	private constructor (props: UserAggregateProps) {
		super(props, UserAggregate.name);
	}

	get email (): EmailValueObject {
		return this.props.email;
	}

	get password (): PasswordValueObject {
		return this.props.password;
	}

	get terms (): TermValueObject[] {
		return this.props.terms;
	}

	get deletedAt (): Date | undefined {
		return this.props.deletedAt;
	}

	deleteAccount (): void {
		this.props.isDeleted = true;
		this.props.deletedAt = new Date();
		this.addDomainEvent(new UserAccountDeletedEvent(this));
	}

	public static create (
		props: UserAggregateProps
	): Result<UserAggregate> {
		return Result.ok<UserAggregate>(new UserAggregate(props));
	}
}

export default UserAggregate;
