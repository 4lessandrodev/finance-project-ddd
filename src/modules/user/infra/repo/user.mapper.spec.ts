import { UserAggregate } from '@modules/user/domain';
import { IpValueObject } from '@modules/user/domain/ip.value-object';
import { TermValueObject } from '@modules/user/domain/term.value-object';
import { DateValueObject, DomainId, EmailValueObject, PasswordValueObject } from 'types-ddd';
import { User } from '../entities/user.schema';
import { UserToDomainMapper } from './user.mapper';

describe('user.mapper', () => {
	//
	const currentDate = new Date('2020-01-01 00:00:00');
	//
	let domain: UserAggregate;
	let persistence: User;
	//
	beforeAll(() => {
		// Create user from domain
		domain = UserAggregate.create(
			{
				ID: DomainId.create('valid_id'),
				email: EmailValueObject.create('valid_mail@domain.com').getResult(),
				password: PasswordValueObject.create('valid_password').getResult(),
				terms: [
					TermValueObject.create({
						acceptedAt: DateValueObject.create(currentDate).getResult(),
						ip: IpValueObject.create('45.192.110.42').getResult(),
						isAccepted: true,
						userAgent: {
							name: 'firefox',
							os: 'LINUX',
							type: 'browser',
							version: '80.0.1',
						},
					}).getResult(),
				],
				createdAt: currentDate,
				updatedAt: currentDate,
			},
		).getResult();

		// Create persistence user
		persistence = {
			createdAt: currentDate,
			email: 'valid_mail@domain.com',
			id: 'valid_id',
			password: 'valid_password',
			isDeleted: false,
			terms: [
				{
					ip: '45.192.110.42',
					acceptedAt: currentDate,
					isAccepted: true,
					userAgent: {
						name: 'firefox',
						os: 'LINUX',
						type: 'browser',
						version: '80.0.1',
					},
				},
			],
			updatedAt: currentDate,
		};
	});
	//
	it('should be defined', () => {
		const mapper = new UserToDomainMapper();
		expect(mapper).toBeDefined();
	});

	it('should convert object from persistence to domain', () => {
		const mapper = new UserToDomainMapper();
		const result = mapper.map(persistence);
		expect(result.getResult()).toEqual(domain);
	});

	it('should convert object from domain to persistence', () => {
		const result = domain.toObject();
		expect(result).toEqual(persistence);
		expect(domain.toObject()).toMatchSnapshot();
	});
});
