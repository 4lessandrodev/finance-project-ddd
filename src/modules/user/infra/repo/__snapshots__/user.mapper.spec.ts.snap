// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user.mapper should convert object from domain to persistence 1`] = `
Object {
  "createdAt": 2020-01-01T00:00:00.000Z,
  "deletedAt": undefined,
  "email": "valid_mail@domain.com",
  "id": "valid_id",
  "isDeleted": false,
  "password": "valid_password",
  "terms": Array [
    Object {
      "acceptedAt": 2020-01-01T00:00:00.000Z,
      "ip": "45.192.110.42",
      "isAccepted": true,
      "userAgent": Object {
        "name": "firefox",
        "os": "LINUX",
        "type": "browser",
        "version": "80.0.1",
      },
    },
  ],
  "updatedAt": 2020-01-01T00:00:00.000Z,
}
`;
